R/generate.R

```diff
@@ -0,0 +1,55 @@
+int_types <- c("integer", "smallint")
+logi_types <- c("boolean")
+num_types <- c("date", "numeric", "real")
+date_types <- c("timestamp without time zone", "timestamp with time zone")
+
+generate_type_check <- function(column_name, data_type, is_nullable) {
+  r_type <- ifelse(data_type %in% int_types, "\"integer\"",
+                   ifelse(data_type %in% logi_types, "\"logical\"",
+                          ifelse(data_type %in% num_types, "\"double\"", "\"character\"")))
+
+  type_check <- paste0("typeof(", column_name, ") == ", r_type)
+  type_check <- ifelse(data_type %in% date_types, paste0("inherits(", column_name, ", \"POSIXct\")"), type_check)
+
+  na_check <- ifelse(is_nullable == "YES", paste0("is.na(", column_name, ") || "), "")
+  paste0("stopifnot(", na_check, type_check, ")")
+}
+
+generate_fake_table_function <- function(table) {
+  ret <- paste0("fake_", table$table_name[[1]], " <- function(")
+  args <- paste0(table$column_name, ifelse(table$is_nullable == "YES", " = NA", ""), collapse = ", ")
+  dat <- paste(table$column_name, "=", table$column_name, collapse = ", ")
+  ret <- paste0(ret, args, ") {\n ")
+  type_checks <- paste0(table$type_check, collapse = "\n  ")
+  ret <- paste(ret, type_checks)
+  ret <- paste0(ret, "\n  data.frame(", dat, ")\n}")
+  ret
+}
+
+generate_functions <- function(tables) {
+  tables$type_check <- generate_type_check(tables$column_name, tables$data_type, tables$is_nullable)
+  tables <- split(tables, tables$table_name)
+  sapply(tables, generate_fake_table_function)
```

Comment:
Please avoid sapply, it leads to some terrible problems.  Here you'd be better off with `vapply(tables, generate_fake_table_function, character(1))` which we usually end up with as a little helper

```
vcapply <- function(X, FUN, ...) {
  vapply(X, FUN, character(1), ...)
}
```